### PR TITLE
CLOUD-651 fix: smartupdate starts after pod deletion

### PR DESCRIPTION
**Pull request by pooknull**
[CLOUD-651](https://jira.percona.com/browse/CLOUD-651) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

https://jira.percona.com/browse/CLOUD-651

Fix statefulSet change detection logic

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| scheduled-backup | passed |
| recreate | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| big-data | passed |
| storage | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| security-context | passed |
| self-healing-advanced | passed |
| self-healing-advanced-chaos | passed |
| limits | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| smart-update | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| pitr | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/1040/commits/aed10cb0dfd4a75fbb6dbf370d21e78c8f8d3dc3
image: `perconalab/percona-xtradb-cluster-operator:PR-1040-aed10cb0`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| big-data | passed |
| haproxy | passed |
| restore-to-encrypted-cluster | passed |
| init-deploy | passed |
| demand-backup | passed |
| limits | passed |
| pitr | passed |
| monitoring-2-0 | passed |
| demand-backup-encrypted-with-tls | passed |
| affinity | passed |
| scaling | passed |
| storage | passed |
| upgrade-haproxy | passed |
| scheduled-backup | passed |
| one-pod | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| auto-tuning | passed |
| security-context | passed |
| self-healing-chaos | passed |
| proxysql-sidecar-res-limits | passed |
| self-healing-advanced | passed |
| users | passed |
| self-healing-advanced-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/1040/commits/aed10cb0dfd4a75fbb6dbf370d21e78c8f8d3dc3
image: `perconalab/percona-xtradb-cluster-operator:PR-1040-aed10cb0`

**Review by egegunes — APPROVED** [review]

File: pkg/controller/pxc/upgrade.go [hunk 1]
```diff
@@ -240,7 +240,20 @@ func (r *ReconcilePerconaXtraDBCluster) smartUpdate(sfs api.StatefulApp, cr *api
...
+		&list,
+		&client.ListOptions{
+			Namespace:     currentSet.Namespace,
+			LabelSelector: labels.SelectorFromSet(sfs.Labels()),
+		},
+	); err != nil {
+		return errors.Wrap(err, "get pod list")
+	}
+	for _, pod := range list.Items {
+		if pod.ObjectMeta.Labels["controller-revision-hash"] != currentSet.Status.UpdateRevision {
```
Comment (delgod) [inline pkg/controller/pxc/upgrade.go:254]: what if the Operator pod was killed in the middle of SmartUpdate.
will Operator continue the SmartUpdate of will be stuck?

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/controller/pxc/upgrade.go [hunk 2]
```diff
@@ -240,7 +240,24 @@ func (r *ReconcilePerconaXtraDBCluster) smartUpdate(sfs api.StatefulApp, cr *api
...
+	if err := r.client.List(context.TODO(),
+		&list,
+		&client.ListOptions{
+			Namespace:     currentSet.Namespace,
+			LabelSelector: labels.SelectorFromSet(sfs.Labels()),
+		},
+	); err != nil {
+		return errors.Wrap(err, "get pod list")
+	}
+	update := false
```
Comment (delgod) [inline pkg/controller/pxc/upgrade.go:253]: can we please rename var? for better code reading
```suggestion
	updateInProgress := false
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scheduled-backup | passed |
| upgrade-haproxy | passed |
| storage | passed |
| recreate | passed |
| haproxy | passed |
| scaling | passed |
| big-data | failed |
| self-healing | passed |
| self-healing-chaos | failed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| self-healing-advanced | passed |
| self-healing-advanced-chaos | failed |
| operator-self-healing | passed |
| operator-self-healing-chaos | failed |
| demand-backup | passed |
| smart-update | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| pitr | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/1040/commits/399ac8e0a58bdeabd08706d9dea6976a8853a0b8
image: `perconalab/percona-xtradb-cluster-operator:PR-1040-399ac8e0`

**Review by egegunes — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| upgrade-haproxy | passed |
| storage | passed |
| proxysql-sidecar-res-limits | passed |
| scaling | passed |
| recreate | passed |
| self-healing | passed |
| users | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| self-healing-chaos | passed |
| tls-issue-self | passed |
| security-context | passed |
| demand-backup | passed |
| tls-issue-cert-manager | passed |
| big-data | passed |
| pitr | passed |
| scheduled-backup | passed |
| tls-issue-cert-manager-ref | passed |
| demand-backup-encrypted-with-tls | passed |
| validation-hook | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| self-healing-advanced | passed |
| self-healing-advanced-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/1040/commits/399ac8e0a58bdeabd08706d9dea6976a8853a0b8
image: `perconalab/percona-xtradb-cluster-operator:PR-1040-399ac8e0`

**Pull request merged**